File: Graph/Himadri/detect-cycle-in-an-undirected-graph.cc
// { Driver Code Starts
#include<bits/stdc++.h>

using namespace std;


 // } Driver Code Ends
/* This function is used to detect a cycle in undirected graph
*  adj[]: array of vectors to represent graph
*  V: number of vertices
*/

bool dfs(int node, bool visited[], int parent, vector<int> adj[]) {
    visited[node] = true;
    for(vector<int>:: iterator itr=adj[node].begin(); itr!=adj[node].end(); itr++) {
        if(!visited[*itr]) {
            if(dfs(*itr, visited, node, adj)) return true;
        }
        else if(*itr!=parent) return true;
    }
    return false;
}

bool isCyclic(vector<int> adj[], int V)
{
   // Your code here
   bool visited[V] = {false};
   for(int node=0; node<V; node++){
       if(!visited[node]) {
           if(dfs(node, visited, -1, adj)) return true;
       }
   }
   return false;

}

// { Driver Code Starts.

int main()
{
    int T;
    cin>>T;
    while(T--)
    {
        int V, E;
        cin>>V>>E;

        // array of vectors to represent graph
        vector<int> adj[V];

        int u, v;
        for(int i=0;i<E;i++)
        {
            cin>>u>>v;

            // adding edge to the graph
            adj[u].push_back(v);
            adj[v].push_back(u);
        }

        cout << isCyclic(adj, V)<<endl;

    }
}
  // } Driver Code Ends

File: Graph/Harshit/Detect_cycle_in_an_undirected_graph.java
// { Driver Code Starts
//Initial Template for Java

import java.util.*;
import java.io.*;
import java.lang.*;

class DriverClass
{
    public static void main (String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        
        while(t-- > 0)
        {
            ArrayList<ArrayList<Integer>> list = new ArrayList<>();
            int nov = sc.nextInt();
            int edg = sc.nextInt();
            
            for(int i = 0; i < nov+1; i++)
                list.add(i, new ArrayList<Integer>());
                
            for(int i = 1; i <= edg; i++)
            {
                int u = sc.nextInt();
                int v = sc.nextInt();
                list.get(u).add(v);
                list.get(v).add(u);
            }
            if(new DetectCycle().isCyclic(list, nov) == true)
                System.out.println("1");
            else System.out.println("0");
        }
    }
}// } Driver Code Ends
//User function Template for Java

/*
ArrayList<ArrayList<Integer>> list: represent graph containing 'V' number of
                                    vertices and edges between them
V: represent number of vertices
*/
class DetectCycle
{
    
    static boolean is_cyclic(ArrayList<ArrayList<Integer>> list,int i,boolean[] vis,int parent)
    {
        vis[i]=true;
        
        ArrayList<Integer> al=list.get(i);
        
        for(int x:al)
        {
            if(!vis[x])   
           { if(is_cyclic(list,x,vis,i))
            {
                return true;
            }
           }
            else if(x!=parent)
            {
                return true;
            }
            
            
        }
        
        return false;
        
        
        
       
    }
    
    
    
    
    
    
    static boolean isCyclic(ArrayList<ArrayList<Integer>> list, int V)
    {
        // int i=0;
        // boolean flag=false;
        // for(ArrayList<Integer> al:list)
        // {
        //     boolean[] vis=new boolean[list.size()];
            
        //     for(int j=0;j<vis.length;j++)
        //     {
        //         vis[j]=false;
        //     }
            
        //   flag=is_it(list,i,i,vis);
        //   if(flag)
        //   {
        //       return true;
        //   }
            
        // }
           boolean[] vis=new boolean[V];
        for(int i=0;i<V;i++)
        {
            if(!vis[i]){
            if(is_cyclic(list,i,vis,-1))
            {
                return true;
            }
            }
        }
        
        return false;
        
        
       // add your code here
    }
}